win32/gui.py

```diff
@@ -713,15 +715,67 @@ def check_yaml_syntax(content):
         warning_popup("Unable to parse yaml: \n %s" % str(e))
         raise
 
+def restartServiceWithDeps(serviceName):
+    # stop the service recursively, looking for dependent services.
+    # if it fails, attempt to restart the ones that were stopped
+    # (so that things are left in the correct state)
+    stopped_deps = list()
+    hscm = win32service.OpenSCManager(None,None,win32service.SC_MANAGER_ALL_ACCESS)
+    try:
+        deps = win32serviceutil.__FindSvcDeps(serviceName)
+
+        for dep in deps:
+            hs = win32service.OpenService(hscm, dep, win32service.SERVICE_ALL_ACCESS)
+            try:
+                win32serviceutil.__StopServiceWithTimeout(hs)
+                stopped_deps.insert(0, dep)
+            except pywintypes.error as exc:
+                if exc.winerror != winerror.ERROR_SERVICE_NOT_ACTIVE:
+                    # walk through and start all the services that have been stopped
+                    for restart_dep in stopped_deps:
+                        try:
+                            win32serviceutil.StartService(restart_dep)
+                        except:
+                            # this is best effort, just continue restarting
+                            pass
+                    raise
+            finally:
+                win32service.CloseServiceHandle(hs)
+
+        # all the deps are stopped.  Stop the one we really wanted to stop
+        hs = win32service.OpenService(hscm, serviceName, win32service.SERVICE_ALL_ACCESS)
+        try:
+            win32serviceutil.__StopServiceWithTimeout(hs)
+        except:
+            # if this fails to stop, then restart the dependent services we
+            # already stopped.
+            for restart_dep in stopped_deps:
+                try:
+                    win32serviceutil.StartService(restart_dep)
+                except:
+                    # this is best effort, just continue restarting
+                    pass
+                raise
```

Comment:
the `raise` here should be one indent lower (i.e. outside the `for` loop)

Reply:
Good catch.

Reply:
I've been able to test the mainline code, starting and stopping and restarting dd-agent with and without trace agent installed.  Haven't been able to test the edge cases, as the services all start and stop the way they're supposed to